### PR TITLE
feat: Appium GitHub workflow scripts

### PR DESCRIPTION
**Pull request by igorgiumellizup**
### Related Issues

#1313

### Description and Example

CI Appium Android \ iOS tests scripts. It might need some tweaks, especially because local tests behave differently from git. Main implementation:
- [x] iOS Appium script 
- [x] Android Appium script (uses BrowserStack)
- [x] Unification script that allows triggering both Android and iOS Appium tests

"Bonus" implementation:
 - Added support for [BrowserStack](https://www.browserstack.com/) testing
 - Improvement and update of core classes in Appium project
 - Organization of Horusec command-line params
 
Some automation scripts ended up not being used, but I decided to keep them anyways because they serve as an important learning reference when creating new scripts (ex setup Android emulator and AVD).  

### Checklist

Please, check if these important points are met using `[x]`:

- [x] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [x] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [x] I am willing to follow-up on review comments in a timely manner.

[PR Guide]: https://github.com/ZupIT/beagle/blob/master/doc/contributing/pull_requests.md

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The title of PR not follow the rules of: https://www.conventionalcommits.org/en/v1.0.0/</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by carlossteinzup — DISMISSED** [review]

File: .github/workflows/automated_tests_appium_ios.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            -   name: Config and run Appium server
+                run: bash fastlane/automatedTests/appium/config_and_run_appium_server.sh
+
+            -   name: Build app-ios project to generate the app file
+                run: | 
+                     set -e
+                     cd tests/appium/app-ios/AppiumApp
+                     xcodebuild -showsdks
+                     xcodebuild -sdk iphonesimulator14.2   
+                     cd $GITHUB_WORKSPACE
```
Comment (theffc) [inline .github/workflows/automated_tests_appium_ios.yml:83]: this change will make it possible to access all dependencies that `AppiumApp` target needs through the Xcode workspace.
```suggestion
                run: | 
                     set -e
                     xcodebuild -showsdks
                     xcodebuild  -workspace iOS/Beagle.xcworkspace -scheme AppiumApp -destination 'platform=iOS Simulator,name=iPhone 11' -quiet
```

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The title of PR not follow the rules of: https://www.conventionalcommits.org/en/v1.0.0/</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/automatedTests/appium/android/run_appium_android_tests.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        "$ANDROID_SDK_ROOT"/platform-tools/adb -s "$line" emu kill
+    done
+}
+
+trap exit SIGHUP SIGINT
+trap cleanup EXIT
+
+echo "#### Starting Appium tests ..."
+chmod +x $APPIUM_PROJECT_DIR/gradlew
+#$APPIUM_PROJECT_DIR/gradlew -p $APPIUM_PROJECT_DIR cucumber -Dplatform=android
```
Comment (matheusribeirozup) [inline fastlane/automatedTests/appium/android/run_appium_android_tests.sh:35]: Is this comment necessary?

File: fastlane/automatedTests/appium/android/run_appium_android_tests.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+chmod +x $APPIUM_PROJECT_DIR/gradlew
+#$APPIUM_PROJECT_DIR/gradlew -p $APPIUM_PROJECT_DIR cucumber -Dplatform=android 
+if $APPIUM_PROJECT_DIR/gradlew -p $APPIUM_PROJECT_DIR cucumber -Dplatform=android; then
+  echo "Gradle task succeeded" >&2
+else
+  echo "Gradle task failed" >&2
+  exit 1
+fi
+
+#echo "Finish!"
```
Comment (matheusribeirozup) [inline fastlane/automatedTests/appium/android/run_appium_android_tests.sh:43]: Is this comment necessary?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/TextStyleScreenSteps.kt [hunk 4]
```diff
@@ -32,6 +32,7 @@ class TextStyleScreenSteps: AbstractStep() {
     @Given("^the Beagle application did launch with the texts on the screen$")
     fun checkScreen() {
         waitForElementWithTextToBeClickable("TextScreen", false, false)
+        //registerCurrentScreenInDatabase("TextStyleScreen01")
```
Comment (matheusribeirozup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/TextStyleScreenSteps.kt:35]: Is this comment necessary? If so, what do you think about providing some information?

**Review by matheusribeirozup — APPROVED** [review]

File: .github/workflows/automated_tests_appium_android.yml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+name: Beagle Appium Android Tests
+
+on:
+  workflow_run:
+    workflows: ["Beagle Appium Tests"] #automated_tests_appium_mobile.yml
```
Comment (uziasferreirazup) [inline .github/workflows/automated_tests_appium_android.yml:21]:
```suggestion
    workflows: ["Beagle Appium Tests"]
```

Screen Shot 2021-02-10 at 15 30 43

I believe that we can remove this comment, because we got to know which pedino file to edit it.

File: .github/workflows/automated_tests_appium_android.yml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+#  pull_request:
+#    branches:
+#      - master
+#    paths:
+#      - 'android/**'
+#      - 'backend/**'
+#      - 'test/appium/app-android/**'
+#      - 'test/appium/project/**'
+#      - '.github/**'
+#      - 'fastlane/automatedTests/appium/**'
```
Comment (uziasferreirazup) [inline .github/workflows/automated_tests_appium_android.yml:45]: is there any reason to keep all this commented out?
I think that if it is not used, we can remove it, this type of configuration is easy to add later and needs to be discussed, it is worth scheduling a meeting to define when we are going to trigger this workflow and which file it will look

File: .github/workflows/automated_tests_appium_android.yml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+jobs:
+  bff_config_job:
+    name: BFF configuration job
+    runs-on: ubuntu-20.04
```
Comment (uziasferreirazup) [inline .github/workflows/automated_tests_appium_android.yml:50]: What do you think about changing the machine to mac, to keep the pattern of everything on our ci / cd running on mac?

File: .github/workflows/automated_tests_appium_android.yml [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      - name: Upload .apk file in BrowserStack and run the Appium test suite
+        run: bash fastlane/automatedTests/appium/android/browserstack/setup_apk_in_browserstack_and_run_android_tests.sh
+
+      - name: Expose failed test screenshot files
+        if: failure()
+        uses: actions/upload-artifact@v2
+        with:
+          name: failed_tests_screenshots
+          path: tests/appium/project/build/screenshots/
+
```
Comment (uziasferreirazup) [inline .github/workflows/automated_tests_appium_android.yml:127]: Is there any reason for not writing these steps and the flow in fastlane? which is our centralized task, if we need to change CI / CD we will not have organizational work, we just need to run the task

File: .github/workflows/horusec_pr.yml [hunk 9]
```diff
@@ -11,13 +11,13 @@ jobs:
...
+      - name: Check out code
+        uses: actions/checkout@v2
+      - name: Running Horusec Security
+        #we need to ignore sample, automatedTests, automated-tests and .gradle because they are not being shipped the Beagle sdk
+        #The two hash was added because we have False Positive on TextInput.swift, we have an enum called PASSWORD, and UrlBuilderDefault.kt, we have a base64 failure but we don't use the base64 on this code
+        env:
+          HORUSEC_CLI_REPOSITORY_AUTHORIZATION: ${{ secrets.HORUSEC_CLI_REPOSITORY_AUTHORIZATION }}
+        run: |
+          curl -fsSL https://horusec.io/bin/install.sh | bash
+          horusec start -p ./ --ignore="**/sample/**,**/.gradle/**,**/automatedTests/**,**/automated-tests/**,**/iOS/Libraries/**,**/iOS/Tests/**,**/flutter/beagle/assets/**,**/tests/appium/**,**/**/gradlew.bat,**/flutter/tests.sh" -a "$HORUSEC_CLI_REPOSITORY_AUTHORIZATION" -n "beagle" -e "true" -u "https://api-horusec.zup.com.br"  -G "true" -F "ba3a605f761ba4d5755128a8d87be8b9fd6e47b4ff3c0d0feba86bfe937e581e,ee5c797510e940f30b021f481137f5df140b2cd3d66d6a15eef07a25c0580f92, 9ebcedfab07c79c309b7e2f1112434c0641819fd6434bf413d8251f7a4698298"
```
Comment (uziasferreirazup) [inline .github/workflows/horusec_pr.yml:23]: Is there any reason to change the spacing that existed on the master?

File: fastlane/automatedTests/appium/android/browserstack/setup_apk_in_browserstack_and_run_android_tests.sh [hunk 10]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
```
Comment (uziasferreirazup) [inline fastlane/automatedTests/appium/android/browserstack/setup_apk_in_browserstack_and_run_android_tests.sh:1]: there is a plugin for fastlane from browserstack https://github.com/browserstack/browserstack-fastlane-plugin,
we really have to use this script, can't we use the plugin?

File: fastlane/automatedTests/appium/android/config_and_run_android_emulator.sh [hunk 11]
```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
```
Comment (uziasferreirazup) [inline fastlane/automatedTests/appium/android/config_and_run_android_emulator.sh:1]: this file is not being used anywhere, is there any reason for it to exist?

File: fastlane/automatedTests/appium/android/install_apk_in_emulator.sh [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
```
Comment (uziasferreirazup) [inline fastlane/automatedTests/appium/android/install_apk_in_emulator.sh:1]: this file is not being used anywhere, is there any reason for it to exist?

File: fastlane/automatedTests/appium/android/run_appium_android_tests.sh [hunk 13]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (uziasferreirazup) [inline fastlane/automatedTests/appium/android/run_appium_android_tests.sh:1]: this file is not being used anywhere, is there any reason for it to exist?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/AbstractStep.kt [hunk 14]
```diff
@@ -93,15 +91,15 @@ abstract class AbstractStep {
...
+        elementValue: String,
+        likeSearch: Boolean,
+        ignoreCase: Boolean
     ): MobileElement {
         val xpath: By = getSearchByValueXpath(elementValue, likeSearch, ignoreCase)
         return AppiumUtil.waitForElementToBePresent(
-                getDriver(),
-                xpath,
-                DEFAULT_ELEMENT_WAIT_TIME_IN_MILL
+            getDriver(),
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/AbstractStep.kt:100]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/Runner.kt [hunk 15]
```diff
@@ -30,7 +30,7 @@ import org.junit.runner.RunWith
      * running all tests on Android: "@android and not @inProgress"
      * running all tests on iOS: "@ios and not @inProgress"
      */
-    tags = "@textinput and not @inProgress",
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/Runner.kt:33]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/TextStyleScreenSteps.kt [hunk 16]
```diff
@@ -21,7 +21,7 @@ import io.cucumber.java.en.Given
 import io.cucumber.java.en.Then
 import org.junit.Assert
 
-class TextStyleScreenSteps: AbstractStep() {
+class TextStyleScreenSteps : AbstractStep() {
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/TextStyleScreenSteps.kt:24]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/AbstractScreen.kt [hunk 17]
```diff
@@ -35,28 +35,8 @@ abstract class AbstractScreen(mobileDriver: AppiumDriver<*>?) {
         PageFactory.initElements(
             AppiumFieldDecorator(
                 driver,
-                Duration.ofMillis(DEFAULT_ELEMENT_WAIT_TIME_IN_MILL)
+                Duration.ofMillis(DEFAULT_SCREEN_WAIT_TIME_IN_MILL)
             ), this
         )
-        waitForScreenToLoad<AbstractScreen>()
     }
-
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/AbstractScreen.kt:43]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/MainScreen.kt [hunk 18]
```diff
@@ -34,22 +34,17 @@ import io.appium.java_client.pagefactory.iOSXCUITFindBy
 class MainScreen(mobileDriver: AppiumDriver<*>?) : AbstractScreen(mobileDriver) {
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/MainScreen.kt:34]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/setup/AppiumConstants.kt [hunk 19]
```diff
@@ -18,5 +18,5 @@ package br.com.zup.beagle.setup
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/setup/AppiumConstants.kt:18]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/AppiumUtil.kt [hunk 20]
```diff
@@ -47,39 +47,25 @@ object AppiumUtil {
     @Synchronized
     fun scrollToElement(
         driver: MobileDriver<*>,
-        element: MobileElement,
-        direction: SwipeDirection,
-        timeoutInMilliseconds: Long
+        elementLocator: By,
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/AppiumUtil.kt:50]: Wouldn't it be better to separate code refactoring into another PR?

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/ImageUtil.kt [hunk 21]
```diff
@@ -19,14 +19,15 @@ package br.com.zup.beagle.utils
...
+import com.github.romankh3.image.comparison.model.ImageComparisonState
 import java.awt.image.BufferedImage
 import java.io.File
 
 
 object ImageUtil {
 
     /**
-     * If there is any difference between the images, writes the image result in file @param resultImageFile
+     * If there is any difference between the images, writes the result image in file @param resultImageFile
```
Comment (uziasferreirazup) [inline tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/ImageUtil.kt:30]: Wouldn't it be better to separate code refactoring into another PR?

**Review by uziasferreirazup — DISMISSED** [review]

File: .github/workflows/automated_tests_appium_android.yml (hunk 7 above)
Comment (igorgiumellizup) [reply .github/workflows/automated_tests_appium_android.yml:50]: The AWS configuration routine and the test run routine are very different routines that need specific OS resources. AWS scripts usually run under Linux. I think including both routines in one job, to use the same OS, would be an anti-pattern approach.

File: .github/workflows/automated_tests_appium_android.yml (hunk 8 above)
Comment (igorgiumellizup) [reply .github/workflows/automated_tests_appium_android.yml:127]: Converting everything to Fastlane would be too much work for now, and it would invalidade all the code in this PR. I suggest we open an issue as an improvement of these scripts, since the ones we have here are working. We can talk this through in a call

File: .github/workflows/horusec_pr.yml (hunk 9 above)
Comment (igorgiumellizup) [reply .github/workflows/horusec_pr.yml:23]: I copy\pasted from the Master repository (https://raw.githubusercontent.com/ZupIT/beagle/master/.github/workflows/horusec_pr.yml). Maybe it's a Windows thing?

File: fastlane/automatedTests/appium/android/browserstack/setup_apk_in_browserstack_and_run_android_tests.sh (hunk 10 above)
Comment (igorgiumellizup) [reply fastlane/automatedTests/appium/android/browserstack/setup_apk_in_browserstack_and_run_android_tests.sh:1]: I think we might consider using this plugin in case we decide to implement a Fastlane version of these scripts.

File: fastlane/automatedTests/appium/android/config_and_run_android_emulator.sh (hunk 11 above)
Comment (igorgiumellizup) [reply fastlane/automatedTests/appium/android/config_and_run_android_emulator.sh:1]: As mentioned in the PR description:
```Some automation scripts ended up not being used, but I decided to keep them anyways because they serve as an important learning reference when creating new scripts (ex setup Android emulator and AVD).```
If everyone agrees, we can remove it.

File: fastlane/automatedTests/appium/android/install_apk_in_emulator.sh (hunk 12 above)
Comment (igorgiumellizup) [reply fastlane/automatedTests/appium/android/install_apk_in_emulator.sh:1]: answered above

File: fastlane/automatedTests/appium/android/run_appium_android_tests.sh (hunk 13 above)
Comment (igorgiumellizup) [reply fastlane/automatedTests/appium/android/run_appium_android_tests.sh:1]: answered above

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/AbstractStep.kt (hunk 14 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/cucumber/steps/AbstractStep.kt:100]: Is this case, it's a performance improvement needed for the test scripts. I considered it related to the PR.

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/AbstractScreen.kt (hunk 17 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/AbstractScreen.kt:43]: answered already. This class wen through performance improvements necessary for the test scripts

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/MainScreen.kt (hunk 18 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/screens/MainScreen.kt:34]: This class wen through performance improvements necessary for the test scripts

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/setup/AppiumConstants.kt (hunk 19 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/setup/AppiumConstants.kt:18]: This class went through performance improvements necessary for the test scripts

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/AppiumUtil.kt (hunk 20 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/AppiumUtil.kt:50]: This class went through performance improvements necessary for the test scripts

File: tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/ImageUtil.kt (hunk 21 above)
Comment (igorgiumellizup) [reply tests/appium/project/src/test/kotlin/br/com/zup/beagle/utils/ImageUtil.kt:30]: It's a small adjustment, not an overall refactoring... I think we can keep it

File: .github/workflows/automated_tests_appium_mobile.yml [hunk 22]
```diff
@@ -0,0 +1,27 @@
+#
```
Comment (theffc) [inline .github/workflows/automated_tests_appium_mobile.yml:1]: Maybe this could be removed, or I just didn't understand it

**Review by theffc — APPROVED** [review]

**Review by matheusribeirozup — APPROVED** [review]

**Review by hernandazevedozup — APPROVED** [review]
As we agreed, We cab approve this PR with the condition of starting a issue to replace the implementation of the steps from github actions to fastlane.

**Review by theffc — APPROVED** [review]
